### DISCUSSION TITLE
Add npm package.json support

### DISCUSSION THREAD

**Discussion by RTann** [quay/claircore#830] [Design]
# Background

ACS's scanner supports scanning npm [package.json](https://docs.npmjs.com/cli/v9/configuring-npm/package-json) files for vulnerabilities. ACS is looking to add this ability to ClairCore.

ClairCore now has the ability to obtain language-specific vulnerability data from OSV. OSV's database includes npm data, so ClairCore has a datasource for npm packages.

# Proposal

ACS's package.json scanning support is pretty straightforward. Simply look for a regular file called `package.json`. ACS attempts to ensure this file is meant for NodeJS instead of some random file which happens to be called `package.json` by checking if `node_modules` or `nodejs` are in the file path, as well. Once we are sure this is a file in which we are interested, we can decode the file to look for the required fields.

This may look like the following:

```
type packageJSON struct {
	Name    `json:"name"`
	Version `json:"version"`
}

...

var pkg packageJSON
err := json.NewDecoder(reader).Decode(&pkg)
...
```

We may determine the related vulnerabilities by matching the package name, version, and repository (`"https://www.npmjs.com/"`).

**Comment by hdonnay** [comment]
I haven't used node really at all, so this is the package metadata as "installed" on disk, right? As in, it's not a requirements range, but one specific version?

**Reply by RTann** [reply]
Correct, this is the exact installed version. Requirements range would be in the `dependencies` section. `version` is the version of the package in [node-semver](https://github.com/npm/node-semver).

**Comment by RTann** [comment]
Created https://issues.redhat.com/browse/PROJQUAY-5099 to track this request.

**Comment by RTann** [comment]
I started a draft PR to see what the implementation may look like: https://github.com/quay/claircore/pull/834
